New Typeahead Options: Disable Filtering / Sorting, Render Custom Footer UI, Full Theming Support

sghoweri opened the pull request:
## Jira
http://vjira2:8080/browse/BDS-1864

## Summary
Adds the ability to disable filtering and/or disable sorting Typeahead search results in addition to being able to customize how to map up complex search data + customize how these complex results get rendered / selected.

![image](https://user-images.githubusercontent.com/1617209/68808032-f19fbe00-0636-11ea-9fa8-6c40e1480a0c.png)

Furthermore, these updates also add 1st class support for color theming + being able to override the color theme via the new `theme` prop.

## Details
With these updates (code. demo, and doc updates), Typeahead now supports
- Customizing / Overriding Theming via the new `theme` prop
- Disabling Filtering / Sorting via the new `no-sort` and `no-filter` props
- Rendering custom footer content via the new `footerTemplate` function

Additionally while putting together the updated component demos, I tossed together a new `<bolt-component-demo>` helper to make it easy to switch between different themes when testing / demoing components.

## How to test
- Review code and docs changes
- Double check our support browsers are working as expected
- @christophersmith262 or @relimck especially, verify that these new APIs handle the downstream needs by the team

danielamorse commented on docs-site/src/components/component-demo/component-demo.js:
```diff
@@ -0,0 +1,152 @@
+ * @return {string}
+ */
+// @todo: move to Bolt Element's new utils file
+const uuid = function b(a) {
```
We have this already, If it works for you: https://github.com/boltdesignsystem/bolt/blob/master/packages/core/utils/get-unique-id.js

sghoweri replied:
🤦‍♂ I could have sworn we did - thanks!

sghoweri replied:
All set!

danielamorse commented on docs-site/src/components/component-demo/component-demo.js:
```diff
@@ -0,0 +1,152 @@
+    : ([1e7] + -1e3 + -4e3 + -8e3 + -1e11).replace(/[018]/g, b);
+};
+
+@customElement('bolt-component-demo')
```
Are you planning to build upon this demo component? I'm wondering because It's a pretty generic name. If not, maybe it should be named something more descriptive, like `bolt-theme-demo`. Thoughts?

sghoweri replied:
I was literally thinking the same thing (thinking things through post PR). How does `<bolt-theme-switcher>` sound?

Related thought: I’ve been thinking a lot about how we might be able to have tiny behavior / context provider components that could “snap together”...

Probably worth an in-person discussion but the premise would be to basically have some tiny wrapper components that provide a specific bit of data or customization (`<bolt-theme-provider>`, `<bolt-spacing-provider>`, `<bolt-resize-provider>`, etc) that could get combined in some higher level component (ex. `<bolt-component-demo>`) which handles the actual placement + rendering controls for the different layers added.

sghoweri replied:
^ back to your original question: for now I’d vote for renaming this to be theme specific + sketch out how we’d handle this for other demo needs.

danielamorse replied:
Thanks for renaming. Works for me! Let's chat about the provider component in-person :)

danielamorse commented on packages/build-tools/package.json:
```diff
@@ -19,7 +19,7 @@
   ],
   "scripts": {
-    "test": "FORCE_COLOR=1 npx lerna exec --scope @bolt/build-tools--test* --concurrency 1 -- npm run test"
+    "test": "FORCE_COLOR=1 npx lerna exec --scope @bolt/build-tools--test* --concurrency 1 -- yarn run test"
```
Why switch to `yarn` here?

sghoweri replied:
Yarn makes it much easier to pass down CLI arguments. Updating these build tool tests is as easy as running `yarn test -u` in the `packages/build-tools` folder!

danielamorse commented on packages/components/bolt-typeahead/__demos__/disable-search-filtering/typeahead.disable-search-filtering.js:
```diff
@@ -0,0 +1,59 @@
+        });
+      });
+
+      // disableSearchFilteringTypeaheadDemo.on(
```
Is this a part of the demo or can it be removed?

sghoweri replied:
Related to my comment below - will remove later today

sghoweri replied:
All set!

sghoweri replied:
All set!

danielamorse commented on packages/components/bolt-typeahead/__demos__/disable-search-filtering/typeahead.disable-search-filtering.twig:
```diff
@@ -0,0 +1,19 @@
+{# {% include "@bolt-components-form/form-textarea.twig" with {
```
Is this a part of the demo or can it be removed?

sghoweri replied:
Will get this cleaned up later today. 

I was originally going to display the raw results data as nicely formatted JSON in the read only text area you flagged above (which I was able to do....). Unfortunately I realized I’d have to separately handle prettying up the stringified JSON so mixed that idea at the last minute 🙃

danielamorse commented on packages/components/bolt-typeahead/package.json:
```diff
@@ -34,6 +34,7 @@
     "access": "public"
   },
   "dependencies": {
+    "shallow-equal": "^1.2.0",
```
Just ran into this recently comparing arrays. Nifty!

sghoweri replied:
💯

danielamorse commented on packages/components/bolt-typeahead/typeahead.autosuggest.js:
```diff
@@ -1,46 +1,91 @@
 export const highlightSearchResults = function(item) {
   const resultItem = item;
   resultItem.matches.forEach(matchItem => {
+    // console.log(matchItem);
```
Please remove

sghoweri replied:
👍🏼

sghoweri replied:
Good to go

danielamorse commented on packages/components/bolt-typeahead/typeahead.js:
```diff
@@ -14,21 +15,32 @@ class BoltTypeahead extends withEvents(withLitHtml()) {
     clearInputText: {
       ...props.string,
-      ...{ default: 'Clear search results' },
+      ...{ default: schema.properties.clear_input_text.default },
```
I like this approach to setting defaults (until we have better schema validation in base class)

danielamorse commented on packages/components/bolt-typeahead/typeahead.scoped.scss:
```diff
@@ -122,65 +184,89 @@ bolt-autosuggest {
     max-height: calc(100vh - #{$bolt-typeahead-button-width});
     max-height: calc(100vh - var(--typeahead-height) );
-  }
+    transition: max-height 0.3s cubic-bezier(0.645,0.045,0.355,1), opacity 0.3s cubic-bezier(0.645,0.045,0.355,1), transform 0.3s cubic-bezier(0.645,0.045,0.355,1);
```
Can we assign this `cubic-bezier` to a variable, even a local one? Also, might be good to document what kind of curve it is.

sghoweri replied:
👍🏼 good catch

sghoweri replied:
I've updated this to use a single variable for the cubic-bezier value itself. Not 100% how best to describe the type of bezier though... 🤔

danielamorse commented on packages/core/styles/index.scss:
```diff
@@ -2,6 +2,7 @@
   BOLT CSS CORE
 \* ------------------------------------ */
 
+@import 'sassy-lists';
```
If we're adding `sassy-lists` could it replace any of our existing [list helpers](https://github.com/boltdesignsystem/bolt/tree/master/packages/core/styles/02-tools)? Doesn't need to happen now, just looking to avoid dupes.

sghoweri replied:
I was hoping for the exact same thing. Ideally we can start to offload some of this custom Sass code and just more heavily use an off the shelf Sass util library. 🤞

sghoweri commented:
> @sghoweri I left some inline comments. None of them blockers, but would like your thoughts.
> 
> I found a couple issues maybe you could take a look at:
> 
> * In IE11 I _can_ click on the Custom Footer button ("See all results"), but in Chrome clicking on it just closes the menu.
> * In IE11, the close icon is missing (or wrong color) on dark theme:
>   ![image](https://user-images.githubusercontent.com/36336/68964502-845f6a80-07a7-11ea-8efb-a5ffccc8d2a8.png)
> 
> Can you reproduce this locally, or is it just me?

Thanks @danielamorse! Looking into these today!

The Chrome behavior is particularly strange... I think I actually ran into that same issue when testing this out initially however I could have sworn I fixed that after changing my approach to rendering footer content... will let you know!

sghoweri commented:
@danielamorse this should be ready to re-review! Just finished testing out these latest updates in IE 11 (outside of Pattern Lab's iframe) and things were looking pretty good all things considered!

adamszalapski commented on docs-site/src/components/theme-switcher/theme-switcher.js:
```diff
@@ -0,0 +1,157 @@
+              class="c-bolt-inline-label c-bolt-inline-label--radio"
+              >xdark</label
+            >
+          </bolt-list-item>
```
Maybe we should generate those items list based on our theme values if they exist somewhere.

sghoweri replied:
@adamszalapski I was actually thinking that as well but held off doing this for a couple reasons:

1. I'm expecting some major revamp work in how we pull in and output core design token data (schema work ties in with this) so I'm a little hesitant to do a lot of data-related work to our theming system until we're working on that.

2. More importantly, I actually don't think this new helper component should always necessarily display every single theme option that's available.

What I'd prefer to see is have this higher order component have every theme option that exists available to _potentially_ display but actually have the component(s) being used inside of it automatically report on what themes they _actually_ support.

For example: If Button only supported `light` and `dark` themes (or some new theme option that other components don't yet support), when the component boots up it could emit an event with data about which themes it supports so this parent component could only show the themes supported.


I definitely want to keep iterating on this — just wanted to get at least something out the door to help with the Typeahead demos for starters!

sghoweri commented:
Just wanted to post a couple of PR description updates to capture the work done in the past couple of days + original work done in this PR.

### Additional Updates
- Upgraded from Preact 8.x to Preact 10
    - Necessary in order to use Preact / React `ref`s to target and customize the Autosuggest React component used under the hood
    - This includes adjusting a few things with the build tools and Babel config, in addition to a few Core and component specific updates
- Significantly improved accessibility support thanks to some super helpful inspiration from https://alphagov.github.io/accessible-autocomplete/examples/. With these updates, we know have audible prompts (that auto-update) for:
    - How many search results have returned
    - Which search result is currently highlighted + the text from the highlighted search result
    - If there aren’t any search results
    - Visually hidden prompt with search instructions
- Under the hood improvements to better manage dropdown behavior
- Switched to using this.$selector syntax for element selectors; this.$parent for the parent selector (inspired by Vue’s API)
- Added MDN’s find-index polyfill for older browsers
